integration-tests/tests/src/tests/sync/mixed.ts

```diff
@@ -15,29 +15,179 @@
-type ValueTester = (actual: Realm.Mixed, inserted: Realm.Mixed) => void | boolean;
+  list!: Realm.List<Realm.Mixed>;
+
+  static schema: ObjectSchema = {
+    name: "MixedClass",
+    properties: {
+      _id: "objectId",
+      value: "mixed",
+      list: "mixed[]",
+    },
+    primaryKey: "_id",
+  };
+}
+
+const bool = true;
+const int = 1;
+const double = 123.456;
+const d128 = BSON.Decimal128.fromString("6.022e23");
+const string = "hello";
+const date = new Date();
+const oid = new BSON.ObjectId();
+const uuid = new BSON.UUID();
+const nullValue = null;
+const data = new Uint8Array([0xd8, 0x21, 0xd6, 0xe8, 0x00, 0x57, 0xbc, 0xb2, 0x6a, 0x15]).buffer;
+
+function getMixedList(realm: Realm) {
+  const expectedObj = { _id: new BSON.ObjectId() };
+  const obj = realm.create(MixedClass, expectedObj);
+
+  const values = [bool, int, double, d128, string, oid, uuid, nullValue, date, data, obj];
+  const expected = [bool, int, double, d128, string, oid, uuid, nullValue, date, data, expectedObj];
+
+  return { values, expected };
+}
+
+function getMixedDict(realm: Realm) {
+  const expectedObj = { _id: new BSON.ObjectId() };
+  const obj = realm.create(MixedClass, expectedObj);
+
+  const values = {
+    bool,
+    int,
+    double,
+    d128,
+    string,
+    oid,
+    uuid,
+    nullValue,
+    date,
+    data,
+    obj,
+  };
+
+  const expected = {
+    bool,
+    int,
+    double,
+    d128,
+    string,
+    oid,
+    uuid,
+    nullValue,
+    date,
+    data,
+    obj: expectedObj,
+  };
+
+  return { values, expected };
+}
+
+function getNestedMixedList(realm: Realm) {
+  const mixList1 = getMixedList(realm);
+  const mixList2 = getMixedList(realm);
+  const mixDict = getMixedDict(realm);
+
+  const values = [...mixList1.values, mixList2.values, mixDict.values];
+  const expected = [...mixList1.expected, mixList2.expected, mixDict.expected];
+
+  return { values, expected };
+}
+
+function getNestedMixedDict(realm: Realm) {
+  const mixDict1 = getMixedDict(realm);
+  const mixDict2 = getMixedDict(realm);
+  const mixList = getMixedList(realm);
+
+  const values = {
+    ...mixDict1.values,
+    innerDict: mixDict2.values,
+    innerList: mixList.values,
+  };
+
+  const expected = {
+    ...mixDict1.expected,
+    innerDict: mixDict2.expected,
+    innerList: mixList.expected,
+  };
+
+  return { values, expected };
+}
+
+function expectRealmList(value: unknown): asserts value is Realm.List<unknown> {
+  expect(value).instanceOf(Realm.List);
+}
+
+function expectRealmDictionary(value: unknown): asserts value is Realm.Dictionary<unknown> {
+  expect(value).instanceOf(Realm.Dictionary);
+}
 
 /**
  * The default tester of values.
  * @param actual The value downloaded from the server.
  * @param inserted The value inserted locally before upload.
  */
-function defaultTester(actual: Realm.Mixed, inserted: Realm.Mixed) {
-  expect(actual).equals(inserted);
+function defaultTester(actual: unknown, inserted: unknown) {
```

Comment:
Since the code infers the expected type from the shape of the inserted POJO, it establishes implicit expectations from the calling code in the behavior of the function. This is especially problematic because `inserted` is `unknown` and therefore accepts any values in combination with the final `else` block, which just compares the string representation. In my opinion it's too easy for the calling code to receive a changed that will render this default tester function useless and as such I fear this will be a source of subtle bugs in the testing code that are hard to discover. Additionally, since this puts restrictions on the shape of the `inserted` values, it's potentially harder for us to test combinations of these and types we add in the future.

Do you see any way of passing the tester function explicitly from the tests?
Although potentially more verbose.

Reply:
I understand your concerns. 
I suppose it depends what you mean by passing the tester function explicitly from the tests. The methods that are used to define the tests accept a tester function, and this is the one that is used when no function is passed. 
If you mean that we should pass a more specific tester function for each test then... We can do it, but it's just much more complicated, because this method not only considers all the possible mixed types (I think at least), but also collections, so it does a recursive comparison if needed.

Reply:
> pass a more specific tester function for each test

Exactly.

> does a recursive comparison if needed

Just an idea on the top of my head: We could implement a higher-order function to create dictionary tester functions with a specific `tester` applied on items.

```typescript
const createDictionaryTester = (itemTester: ValueTester) => (actual: unknown, inserted: unknown) {
  expectRealmDictionary(actual);
  if (typeof inserted !== "object" || inserted === null) {
    throw new Error("Expected an object of values");
  }
  for (const [insertedKey, insertedValue] of Object.entries(inserted)) {
    itemTester(actual[insertedKey], insertedValue);
  }
}
```

To test a nested dictionary `{ foo: { bar: ["baz"] } }` this be chained like:

```typescript
tester: createDictionaryTester(createDictionaryTester(stringTester))
```

Reply:
I think there are a couple of things to consider:
- The values that are being used the tests look something like `[1, true, "string", ...., object, innerList, innerDicti]`. S
Thus the testing methods wouldn't look particularly pretty.
- It gets even more messy in the other PR with additional tests. In that PR I don't have a single mixed value that I'm trying to roundtrip like here, but I do a lot of modifications to the list/dictionary in order to verify it all works correctly, so I can't have one specific testing function. 
Overall I tried to make a testing function that would be generic enough to work with different kind of tests, even though it is not maybe the most specific or strict. I think that if we go the more strict route we either need to simplify the test cases (mostly considering the additional PR) or we need to greatly complicate our tester.

Reply:
🤔 Yeah. I agree, there are tradeoffs here, which complicates the decision.
I wanted to mention my reservations, but I won't object to this merging as is.